### PR TITLE
Fixing VM release-before-retain when the source and target alias.

### PR DESCRIPTION
**Pull request by benvanik**

### CODE REVIEW

File: iree/vm/list.c [hunk 1]
```diff
@@ -505,8 +505,9 @@ IREE_API_EXPORT void* iree_vm_list_get_ref_deref(
   return value.ptr;
 }
 
-IREE_API_EXPORT iree_status_t iree_vm_list_get_ref_assign(
-    const iree_vm_list_t* list, iree_host_size_t i, iree_vm_ref_t* out_value) {
+static iree_status_t iree_vm_list_get_ref_assign_or_retain(
```
Comment (stellaraccident) [inline iree/vm/list.c:508]: For us mortals, would you mind adding a comment to this function describing what it does?

**Review by stellaraccident — APPROVED** [review]

**Review by stellaraccident — APPROVED** [review]

**Comment by soraros** [comment]
yn

**Comment by antiagainst** [comment]
FYI: this causes a crash when executing VMVX as shown in https://buildkite.com/iree/iree-benchmark/builds/1128#57b828f1-f704-432c-a884-07f1c4271e7e.

Steps to reproduce:

1. Download input model from the URL at https://github.com/google/iree/blob/eed0432/benchmarks/TensorFlow/CMakeLists.txt#L42.
2. Generate the artifact with `iree/tools/iree-translate -iree-input-type=mhlo -iree-mlir-to-vm-bytecode-module -iree-hal-target-backends=vmvx -iree-flow-inline-constants-max-byte-length=2048 -o iree.vmfb /path/to/MobileNetV2.mlir`.
3. Run with `iree/tools/iree-benchmark-module --driver=vmvx --module_file=iree.vmfb --entry_function=call --function_input=1x224x224x224xf32 --task_topology_group_count=3`.

This does not need an Android machine. Using `iree-benchmark-module` compiled for x86 can also reproduce the crash.

**Comment by benvanik** [comment]
I think the vmvx issue may be an unrelated bug/leak that this was catching - one of the dispatches is runaway retaining and never releasing: 
image

**Pull request merged**